### DISCUSSION TITLE
Is there an extension out there that wraps the zipalign tool - pack android

### DISCUSSION THREAD

**Discussion by nick5454** [cake-build/cake#3342] [Q&A]
Does anyone know of an extension that installs this tool? I'd rather use an extension that does than install it myself.

When you use jar to combine an .apk (Android), you have to call zipalign afterwards.

And another question what is the best way to install this tool, since it probably won't exist on hosted Macos in Azure Devops. It looks like only through Android Studio.

So if I packed this into my cake private nuget, I just create a tools directory and throw it in?

I see this [here](https://cakebuild.net/docs/writing-builds/tools/installing-tools)

So would I just add the zipalign tool into my tools directory and call this below ( cake is running on the Macos hosted by azure devops)?

'''Task("Install-XUnit")
    .Does(()=> {
    
    StartProcess("./", new ProcessSettings {
        Arguments = new ProcessArgumentBuilder()
            .Append("install")
            .Append("zipalign")
        });
});'''

**Comment by augustoproiete** [comment]
@nick5454 I'm not aware of an existing Cake addin that wraps the zipalign tool - pack android as of this writing. You might want build one or perhaps add a post in the [extension request](https://github.com/cake-build/cake/discussions/categories/extension-requests) to see if there are any takers.

You are on the right track though. You'd use `StartProcess` (or `StartAndReturnProcess` if you want more control) to call this `zipalign` tool from a Cake build, once you know the tool is present on the build node.

The cleanest way to ensure the tool is installed for you, would be (as you said) to package it up into a nuget package, publish to a feed that is accessible from the build node, and use the `#tool` directive to ensure the tool is installed.

Then you would use `Context.Tools.Resolve` to ask Cake to find the tool, and the you can run it.

> So if I packed this into my cake private nuget, I just create a tools directory and throw it in?

Yes, just create a `tools` folder with everything that you need for the tool to run. It's not mandatory for this folder to be called `tools` for it to work with Cake, but it's a very common practice.

You can see an example of this by looking at my [ExcelDnaPack-NuGet](https://github.com/augustoproiete/ExcelDnaPack-NuGet) repo. There I'm packaging up a tool called `ExcelDnaPack.exe` along with other files that it needs, and publishing to the NuGet Gallery as [ExcelDnaPack](https://www.nuget.org/packages/ExcelDnaPack/).

![image](https://user-images.githubusercontent.com/177608/117549586-9d01b700-b011-11eb-9d85-8df7c9439670.png)

To use that package with Cake directly, you'd do something like the following:

```csharp
#tool "nuget:?package=ExcelDnaPack&version=1.1.1"

Task("Pack")
    .Does(context =>
{
    var excelDnaPackExePath = Context.Tools.Resolve("ExcelDnaPack.exe");

    var settings = new ProcessSettings()
        .WithArguments(args => args
            .AppendQuoted(@"MyAddins\FirstAddin.dna")
            .Append("/Y")
        );

    var exitCode = -1;
    using (var process = StartAndReturnProcess(excelDnaPackExePath, settings))
    {
        process.WaitForExit();
        exitCode = process.GetExitCode();
    }

    Information("Exit code: {0}", exitCode);
    // ...
});

RunTarget("Pack");
```

This is how [I initially used this package with Cake](https://github.com/augustoproiete/ExcelDnaPack-NuGet/tree/v1.1.1#using-exceldnapack-with-cake-build). Later I decided to create a dedicated Cake addin that wraps this tool ([Cake.ExcelDnaPack](https://github.com/cake-contrib/Cake.ExcelDnaPack)), which makes it easier to use with specific Cake aliases:

```csharp
#tool "nuget:?package=ExcelDnaPack&version=1.1.1"
#addin "nuget:?package=Cake.ExcelDnaPack&version=1.0.0"

Task("Pack")
    .Does(context =>
{
    ExcelDnaPack(@"MyAddins\FirstAddin.dna");
});

RunTarget("Pack");
```

**Reply by nick5454** [reply]
@augustoproiete Wow awesome reply! Thanks for being so concise. I may actually create a cake extension, because i've actually had to go through hoops getting the android build working.

I actually build the apps in the build pipeline. On release, I delete the configuration and put in the configuration file based on the Environment. But this also means I have to repack and resign on all platforms. This has causes some unique issues to overcome.
